libopae/init_ase.c

```diff
@@ -0,0 +1,79 @@
+// Copyright(c) 2019, Intel Corporation
+//
+// Redistribution  and  use  in source  and  binary  forms,  with  or  without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of  source code  must retain the  above copyright notice,
+//   this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+// * Neither the name  of Intel Corporation  nor the names of its contributors
+//   may be used to  endorse or promote  products derived  from this  software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif // HAVE_CONFIG_H
+
+#include <opae/init.h>
+#include <opae/utils.h>
+#include "pluginmgr.h"
+#include "opae_int.h"
+#undef __USE_GNU
+
+#include "safe_string/safe_string.h"
+#include <stdio.h>
+#include <stdarg.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#ifndef __USE_GNU
+#define __USE_GNU
+#endif // __USE_GNU
+#include <pthread.h>
+
+#define PATH_MAX 1024
+
+__attribute__((constructor)) STATIC void opae_ase_init(void)
+{
+	char cfg_path[PATH_MAX];
+	char *opae_path = getenv("OPAE_PLATFORM_ROOT");
```

Comment:
Shouldn't this be searching the OPAE SDK installation directory (where ASE is stored), not the release directory?

Reply:
@michael-adler Do we have an environment variable for the OPAE SDK installation directory?

Reply:
No, but it is known at build time since it is all inside OPAE.

You can see an example in getASESrcPath() in https://github.com/OPAE/opae-sdk/blob/4a7de2e54c0dd308c00d5386a2736eebd3c16b64/ase/scripts/afu_sim_setup#L46. There has to be a special CMake rule to do the substitution, so a bit messy, but definitely possible.

Reply:
@michael-adler I would put the opae_ase.cfg in the stand search locations for config file.

Reply:
I don't understand what that means.

A goal here should be that this keeps working with old releases, so we don't want to modify anything under OPAE_PLATFORM_ROOT.

On the other hand, your change is part of the OPAE SDK, so you are free to change whatever you need in the SDK sources or in the generated OPAE installation tree.

Reply:
I could be wrong, but from inspection it looks like the new code still isn't searching the installation directory. Aren't you going to create a default opae_ase.cfg and store it in <install>/etc/opae/opae_ase.cfg or (maybe better) under <install>/share/opae/ase? That would leave ASE completely functional with no user intervention.

If I'm reading the code correctly, it still requires a change by the user.

Reply:
Installed opae_ase.cfg to ~/.local/opae now so that ASE can work the same as before plugin changes.

Reply:
~/.local assumes it is the ASE user who did the installation, which is not necessarily true. It won't work for a shared installation. Consider how the DCP test system works. OPAE is installed by a nightly script to some shared location. It is put on a PATH. The vLab is configured the same way.

I don't understand the resistance to putting a default copy of opae_ase.cfg in the place where ASE itself is stored. All our tools have to be able to find <install>/share/ase already. Isn't it the obvious place for a default opae_ase.cfg?

Reply:
It is already installed in the shared location share/opae/ase. I just need to put the shared location in the searching path. Is there an environment variable for the installation path? Probably OPAE_PLATFORM_ROOT can be used.

Reply:
There is no environment variable. The CMake build script knows where it is going to be installed and we use CMake variable substitution to construct sources that compile in the installation path. I don't know if there are C examples, but we do this for Python scripts like afu_sim_setup. Hence the reference to getASESrcPath() in the comments above.

The easiest way might be to create a header file that has the installation directory as a macro, add a CMake rule to do variable substitution on the header file, and put the build directory with the constructed header file on the C header search path.